src/keys.c

```diff
@@ -257,16 +256,34 @@ ece_aes128gcm_derive_key_and_nonce(EC_KEY* recvPrivKey, EC_KEY* senderPubKey,
   ece_buf_t prk;
   ece_buf_reset(&prk);
 
-  err = ece_compute_secret(recvPrivKey, senderPubKey, &sharedSecret);
+  err = ece_compute_secret(localKey, remoteKey, &sharedSecret);
   if (err) {
     goto end;
   }
 
   // The new "aes128gcm" scheme includes the sender and receiver public keys in
   // the info string when deriving the Web Push PRK.
-  err = ece_aes128gcm_generate_info(
-    recvPrivKey, senderPubKey, ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX,
-    ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX_LENGTH, &prkInfo);
+  switch (mode) {
+  case ECE_MODE_ENCRYPT:
+    // For encryption, the remote static public key is the receiver key, and the
+    // local ephemeral private key is the sender key.
+    err = ece_aes128gcm_generate_info(
+      remoteKey, localKey, ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX,
+      ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX_LENGTH, &prkInfo);
+    break;
+
+  case ECE_MODE_DECRYPT:
+    // For decryption, the local static private key is the receiver key, and the
+    // remote ephemeral public key is the sender key.
+    err = ece_aes128gcm_generate_info(
+      localKey, remoteKey, ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX,
+      ECE_AES128GCM_WEB_PUSH_PRK_INFO_PREFIX_LENGTH, &prkInfo);
+    break;
+
+  default:
+    assert(false);
+    err = ECE_ERROR_DECRYPT;
```

Comment:
Audit the other errors in this function to make sure they're not `ENCRYPT` or `DECRYPT` specific.